riscv/src/elf/debug_info.rs

```diff
@@ -0,0 +1,565 @@
+use std::{
+    borrow::Cow,
+    collections::{BTreeMap, BTreeSet, HashMap},
+    path::Path,
+};
+
+use gimli::{
+    read::AttributeValue, DebuggingInformationEntry, Dwarf, EndianSlice, LittleEndian, Operation,
+    Unit, UnitRef,
+};
+use goblin::elf::{
+    sym::{STT_FUNC, STT_OBJECT},
+    Elf, SectionHeader,
+};
+use itertools::Itertools;
+
+use super::AddressMap;
+
+type Reader<'a> = EndianSlice<'a, LittleEndian>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("no debug information available")]
+    NoDebugInfo,
+    #[error("DIE tree traversal skipped a level")]
+    UnexpectedLevel,
+    #[error("failed to parse debug information: {0}")]
+    Parsing(#[from] gimli::Error),
+}
+
+/// Debug information extracted from the ELF file.
+#[derive(Default)]
+pub struct DebugInfo {
+    /// List of source files: (directory, file name).
+    pub file_list: Vec<(String, String)>,
+    /// Relates addresses to source locations.
+    pub source_locations: Vec<SourceLocationInfo>,
+    /// Maps (addresses, disambiguator) to symbol names. The disambiguator is
+    /// used to distinguish between multiple symbols at the same address. (i.e.
+    /// turns BTreeMap into a multimap.)
+    pub symbols: SymbolTable,
+    /// Human readable notes about an address
+    pub notes: HashMap<u32, String>,
+}
+
+#[derive(Debug)]
+pub struct SourceLocationInfo {
+    pub address: u32,
+    pub file: u64,
+    pub line: u64,
+    pub col: u64,
+}
+
+impl DebugInfo {
+    /// Extracts debug information from the ELF file, if available.
+    pub fn new<D>(
+        elf: &Elf,
```

Comment:
this function is a bit big, maybe break the body of this function into a few extra functions? maybe some helpers for each of the fields in `DebugInfo` or sth else that makes sense

Reply:
Done.

Reply:
noice!